src/utils/gauges.ts

```diff
@@ -366,9 +414,347 @@ export async function getGaugeRewardsUserData(
+  const gaugeWorkingBalancesPromise = account
+    ? ethCallProvider.tryAll(
+        gaugeMulticallContracts.map((gaugeContract) =>
+          gaugeContract.working_balances(account),
+        ),
+      )
+    : null
+  const gaugeLpTokenAddressesPromise = ethCallProvider.tryAll(
+    gaugeMulticallContracts.map((gaugeContract) => gaugeContract.lp_token()),
+  )
+  const gaugeNamesPromise = ethCallProvider.tryAll(
+    gaugeMulticallContracts.map((gaugeContract) => gaugeContract.symbol()),
+  )
+  const gaugeKillStatusesPromise = ethCallProvider.tryAll(
+    gaugeMulticallContracts.map((gaugeContract) => gaugeContract.is_killed()),
+  )
+
+  return {
+    gaugeBalancePromise,
+    gaugeWorkingBalancesPromise,
+    gaugeWorkingSuppliesPromise,
+    gaugeTotalSupplyPromise,
+    gaugeLpTokenAddressesPromise,
+    gaugeNamesPromise,
+    gaugeKillStatusesPromise,
+  }
+}
+
+function buildLpTokenToPool(basicPools: BasicPools): Partial<{
+  [lpToken: string]: GaugePool
+}> {
+  return Object.values(basicPools || {}).reduce(
+    (prevData, { lpToken, poolAddress, poolName }) => {
+      return {
+        ...prevData,
+        [lpToken]: { poolAddress, poolName },
+      }
+    },
+    {},
+  )
+}
+
+function buildLpTokenAddressToGauge(
+  chainId: ChainId,
+  lpTokenToPool: Partial<{
+    [lpToken: string]: GaugePool
+  }>,
+  sdlRates: (BigNumber | null)[],
+  gaugeAddresses: string[],
+  gaugeWeights: (BigNumber | null)[],
+  gaugeRelativeWeights: (BigNumber | null)[],
+  gaugeRewards: Awaited<
+    ReturnType<
+      | typeof getGaugeRewardsFromTokens
+      | typeof getGaugeRewardsFromTokensSidechain
+    >
+  >,
+  gaugeRewardTokens: (string | null)[][],
+  gaugeBalances: (BigNumber | null)[] | null,
+  gaugeWorkingBalances: (BigNumber | null)[] | null,
+  gaugeWorkingSupplies: (BigNumber | null)[],
+  gaugeTotalSupplies: (BigNumber | null)[],
+  gaugeLpTokenAddresses: (string | null)[],
+  gaugeNames: (string | null)[],
+  gaugeKillStatuses: (boolean | null)[],
+) {
+  return gaugeAddresses.reduce((previousGaugeData, gaugeAddress, index) => {
+    const lpTokenAddress = gaugeLpTokenAddresses[index]?.toLowerCase()
+    const pool = lpTokenToPool[lpTokenAddress || ""] as GaugePool
+    if (!lpTokenAddress || !pool) return previousGaugeData
+
+    const isValidPoolAddress = Boolean(
+      pool.poolAddress && !isAddressZero(pool.poolAddress),
+    )
+
+    const poolAddress = isValidPoolAddress ? pool.poolAddress : null
+    const sdlRate = sdlRates[index] || Zero
+    const sdlReward = {
+      periodFinish: BN_MSIG_SDL_VEST_END_TIMESTAMP,
+      rate: sdlRate,
+      tokenAddress: SDL_TOKEN_ADDRESSES[chainId].toLowerCase(),
+      isMinter: true,
+    }
+    const gaugeTokens = gaugeRewardTokens[index]
+    const gaugeTokenReward = (gaugeRewards[index] as GaugeTokenRewardData[])
+      .map((reward, tokenIndex) => {
+        if (gaugeTokens[tokenIndex] != null && reward) {
+          const tokenAddress = gaugeTokens[tokenIndex]?.toLowerCase()
```

Comment:
@penandlim Can I get 2nd look at this line? I'm using the `gaugeTokens` list from the `getGaugeRewardTokens` call on L595. I think this is how the data matches up.

Reply:
`gaugeTokens` here means array of external reward token addresses for this gauge right?

LGTM in that case